AndroidSDKTests/src/test/java/com/leanplum/internal/RequestTest.java

```diff
@@ -323,4 +338,30 @@ public void testGetRequestsWithEncodedStringStoredRequests() {
     return requests;
   }
 
+  private static class ThreadWaiter implements Waiter {
+    Instant t1, t2, t3, t4;
+    @Override
+    public void beforeRead() {
+      t1 = Instant.now();
+    }
+
+    @Override
+    public void afterRead() {
+      t2 = Instant.now();
+    }
+
+    @Override
+    public void beforeWrite() {
+      t3 = Instant.now();
```

Comment:
in addition to recording the time beforeWrite, we also need to block on a semaphore here, so that a read can potentially happen in a race condition, then in the test we allow the write to proceed and assert that no race condition occurred.